core/src/main/java/org/apache/calcite/rel/metadata/CyclicMetadataException.java

```diff
@@ -21,15 +21,12 @@
  * computing metadata.
  */
 public class CyclicMetadataException extends RuntimeException {
-  /** Singleton instance. Since this exception is thrown for signaling purposes,
-   * rather than on an actual error, re-using a singleton instance saves the
-   * effort of constructing an exception instance. */
-  @SuppressWarnings("ThrowableInstanceNeverThrown")
-  public static final CyclicMetadataException INSTANCE =
-      new CyclicMetadataException();
-
-  /** Creates a CyclicMetadataException. */
-  private CyclicMetadataException() {
+  /**
+   * Creates a CyclicMetadataException.
+   * Reusing a singleton instance is a bad idea because of wrong stack trace.
+   * See https://stackoverflow.com/questions/15090664
```

Comment:
You can remove these 2 lines. Better to move them into the commit messages, instead of commenting here.

Reply:
this is removed from the code and moved to commit message